### PR TITLE
Magento category synchronizer

### PR DESCRIPTION
**Pull request by adudekyeti**
Magento category synchronizer

### CODE REVIEW

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 1]
```diff
@@ -8,21 +8,536 @@
...
+	 * @param array $categoryYF
+	 * @param array $categoryMagento
+	 *
+	 * @throws \yii\db\Exception
+	 *
+	 * @return int
+	 */
+	public function updateCategoryYF(array $categoryYF, array $categoryMagento): int
+	{
+		$templateId = \Vtiger_Field_Model::getInstance('pscategory', \Vtiger_Module_Model::getInstance('Products'))->getFieldParams();
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Category.php]: add cache - $templateId

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 2]
```diff
@@ -8,21 +8,536 @@
...
+			'state' => '{"loaded":true,"opened":false,"selected":false,"disabled":false}'
+		])->execute();
+		$this->saveCategoryMappingYF($category['id'], $categoryId);
+		if (isset($this->mapCategoryYF[$category['parent_id']])) {
+			$category['parent_id'] = $this->mapCategoryYF[$category['parent_id']];
+		} else {
+			$category['parent_id'] = $categoryId;
+		}
+		$parentId = $this->parseParentYF($category);
+		(new Query())->createCommand()->update('vtiger_trees_templates_data', ['parentTree' => $parentId], ['tree' => 'T' . $categoryId, 'templateid' => $templateId])->execute();
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Category.php]: \App\Db::getInstance()->createCommand()

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 3]
```diff
@@ -8,21 +8,536 @@
...
+	 *
+	 * @throws \yii\db\Exception
+	 *
+	 * @return int
+	 */
+	public function saveCategoryYF(array $category): int
+	{
+		$categoryId = $this->getLastIdYF();
+		$templateId = \Vtiger_Field_Model::getInstance('pscategory', \Vtiger_Module_Model::getInstance('Products'))->getFieldParams();
+		$result = (new Query())->createCommand()->insert('vtiger_trees_templates_data', [
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Category.php]: \App\Db::getInstance()->createCommand()

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 4]
```diff
@@ -8,21 +8,536 @@
...
+	 * Method to return last categories tree id.
+	 *
+	 * @throws \yii\db\Exception
+	 *
+	 * @return int
+	 */
+	public function getLastIdYF(): int
+	{
+		$templateId = \Vtiger_Field_Model::getInstance('pscategory', \Vtiger_Module_Model::getInstance('Products'))->getFieldParams();
+		$categoriesYF = (new Query())
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Category.php]: Duplicate query

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 5]
```diff
@@ -8,21 +8,536 @@
...
 namespace App\Integrations\Magento\Synchronizator;
 
+use App\Db\Query;
+
 /**
  * Category class.
  */
 class Category extends Base
 {
+	public $parentsId = [];
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Category.php]: Add docs

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 6]
```diff
@@ -8,21 +8,583 @@
...
+					}
+				} else {
+					if (!$master) {
+						$this->saveCategoryYF($categoryMagento);
+					} else {
+						$deletedParent = $this->deleteCategoryMagento($categoryMagento);
+					}
+				}
+				if (!$deletedParent && !empty($categoryMagento['children_data'])) {
+					$this->updateCategoriesYF($master, $categoryMagento['children_data'], $this->rootCategory);
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Category.php]: ?

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 7]
```diff
@@ -8,21 +8,583 @@
...
+	{
+		$changed = false;
+		if (
+			$categoryYF['name'] !== $categoryMagento['name'] ||
+			(int) $categoryYF['is_active'] !== (int) $categoryMagento['is_active'] ||
+			$this->hasChangedCategory($categoryYF, $categoryMagento)
+		) {
+			$changed = true;
+		}
+		return $changed;
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Category.php]: ?

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 8]
```diff
@@ -8,21 +8,583 @@
...
+	 *
+	 * @return bool
+	 */
+	public function hasChangedCategory(array $categoryYF, array $categoryMagento)
+	{
+		$changed = false;
+		if (isset($this->mapCategoryMagento[$categoryYF['parent_id']]) && $this->mapCategoryMagento[$categoryYF['parent_id']] !== $categoryMagento['parent_id']) {
+			$changed = true;
+		}
+		return $changed;
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Category.php]: ?

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 9]
```diff
@@ -8,21 +8,583 @@
...
+	/**
+	 * Method to return Magento categories parsed to array.
+	 *
+	 * @return array
+	 */
+	public function getCategoriesMagento()
+	{
+		try {
+			$categoriesList = $this->connector->request('GET', '/rest/all/V1/categories');
+			$this->categoriesMagento['children_data'] = [\json_decode($categoriesList, true)];
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
			$this->categoriesMagento['children_data'] = [\App\Json::decode($categoriesList, true)];
```

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 10]
```diff
@@ -8,21 +8,583 @@
...
+					'name' => $category['name'],
+					'parent_id' => $this->mapCategoryMagento[$category['parent_id']] ?? 0,
+					'is_active' => $category['is_active'],
+					'include_in_menu' => '1',
+					'isActive' => 'true',
+					'level' => $category['level'] + 1,
+				]]), true);
+			$this->saveCategoryMappingYF($categoryRequest['id'], $category['id']);
+			$result = true;
+		} catch (\GuzzleHttp\Exception\ClientException $e) {
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Category.php]: ?

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 11]
```diff
@@ -8,21 +8,577 @@
...
+	 * @throws \yii\db\Exception
+	 *
+	 * @return array
+	 */
+	public function getCategoriesYF(): array
+	{
+		foreach ($this->getTemplatesData() as $categoryYF) {
+			$treeID = (int) str_replace('T', '', $categoryYF['tree']);
+			$parentTreeID = str_replace($categoryYF['tree'], '', $categoryYF['parentTree']);
+			$parentTreeID = array_filter(explode('::', str_replace('T', '', $parentTreeID)));
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]: App\Fields\Tree::getParentIdx ?

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 12]
```diff
@@ -8,21 +8,577 @@
...
+	 * @return array
+	 */
+	public function getCategoriesMagento()
+	{
+		try {
+			$categoriesList = $this->connector->request('GET', '/rest/all/V1/categories');
+			$this->categoriesMagento['children_data'] = [\App\Json::decode($categoriesList)];
+			$this->parseMagentoCategory($this->categoriesMagento);
+		} catch (\Throwable $ex) {
+			\App\Log::error('Error while getting categories from magento: ' . $ex->getMessage());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
			\App\Log::error('Error while getting categories from magento: ' . $ex->getMessage(), 'Integrations/Magento');
```

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 13]
```diff
@@ -8,21 +8,577 @@
...
+					'parent_id' => $this->mapCategoryMagento[$category['parent_id']] ?? 0,
+					'is_active' => $category['is_active'],
+					'include_in_menu' => '1',
+					'isActive' => 'true',
+					'level' => $category['level'] + 1,
+				]]));
+			$this->saveCategoryMappingYF($categoryRequest['id'], $category['id']);
+			$result = true;
+		} catch (\Throwable $ex) {
+			\App\Log::error('Error during saving magento category: ' . $ex->getMessage());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
			\App\Log::error('Error during saving magento category: ' . $ex->getMessage(), 'Integrations/Magento');
```

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 14]
```diff
@@ -8,21 +8,577 @@
...
+		try {
+			$this->connector->request('PUT', '/rest/all/V1/categories/' . $categoryMagento['id'], [
+				'category' => [
+					'name' => $categoryYF['name'],
+					'isActive' => $categoryYF['is_active'],
+				]
+			]);
+			$result = true;
+		} catch (\Throwable $ex) {
+			\App\Log::error('Error during updating magento category: ' . $ex->getMessage());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
			\App\Log::error('Error during updating magento category: ' . $ex->getMessage(), 'Integrations/Magento');
```

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 15]
```diff
@@ -8,21 +8,577 @@
...
+	public function deleteCategoryMagento($categoryMagento): bool
+	{
+		$result = false;
+
+		if (!\in_array($categoryMagento['id'], static::$nonEditable)) {
+			try {
+				$this->connector->request('DELETE', '/rest/all/V1/categories/' . $categoryMagento['id'], []);
+				$result = true;
+			} catch (\Throwable $ex) {
+				\App\Log::error('Error during deleting magento category: ' . $ex->getMessage());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
				\App\Log::error('Error during deleting magento category: ' . $ex->getMessage(), 'Integrations/Magento');
```

File: app/Integrations/Magento/Synchronizator/Category.php [hunk 16]
```diff
@@ -8,21 +8,577 @@
...
+	 *
+	 * @return bool
+	 */
+	public function updateCategoryParentMagento($categoryYF, $categoryMagento): bool
+	{
+		try {
+			$this->connector->request('PUT', '/rest/all/V1/categories/' . $categoryMagento['id'] . '/move', ['parentId' => $this->mapCategoryMagento[$categoryYF['parent_id']] ?? 0]);
+			$result = true;
+		} catch (\Throwable $ex) {
+			\App\Log::error('Error during moving magento category: ' . $ex->getMessage());
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Category.php]:
```suggestion
			\App\Log::error('Error during moving magento category: ' . $ex->getMessage(), 'Integrations/Magento');
```

File: app/Integrations/Magento/Synchronizator/Category.php (hunk 11 above)
Comment (adudekyeti) [reply app/Integrations/Magento/Synchronizator/Category.php]: I can not use it.

**Pull request merged**